File: TripKitUI/views/results/TKUITimePickerSheet.m
//
//  TKUITimePickerSheet.m
//  TripKitUI
//
//  Created by Adrian Schoenig on 24/09/13.
//
//

#import "TKUITimePickerSheet.h"

#ifdef TK_NO_MODULE
#import "TripKit.h"
#import "TripKit/TripKit-Swift.h"
#else
@import TripKit;
#import "TripKitUI/TripKitUI-Swift.h"
#endif

@interface TKUITimePickerSheet ()

@property (nonatomic, assign) BOOL didSetTime;

@property (nonatomic, weak) UISegmentedControl *timeTypeSelector;
@property (nonatomic, weak) UISegmentedControl *doneSelector;

@property (nonatomic, assign) BOOL includeTimeType;

@end

@implementation TKUITimePickerSheet

#pragma mark - UIView

- (instancetype)initWithTime:(nullable NSDate *)time
                    timeType:(TKTimeType)timeType
                    timeZone:(NSTimeZone *)timeZone
{
  self = [super initWithFrame:CGRectMake(0, 0, 320, 216)];
  if (self) {
    self.includeTimeType = YES;
    
    [self commonInit];

    if (! time) {
      time = [NSDate date];
    }
    
    self.timePicker.timeZone = timeZone;
    self.timePicker.date     = time;
    [self setSelectedTimeType:timeType];
  }
  return self;
}

- (instancetype)initWithTime:(nullable NSDate *)time
                    timeZone:(NSTimeZone *)timeZone
{
  self = [super initWithFrame:CGRectMake(0, 0, 320, 216)];
  if (self) {
    self.includeTimeType = NO;
    
    [self commonInit];

    if (! time) {
      time = [NSDate date];
    }
    self.timePicker.datePickerMode = UIDatePickerModeDateAndTime;
    self.timePicker.timeZone = timeZone;
    self.timePicker.date     = time;
  }
  return self;
}

- (instancetype)initWithDate:(nullable NSDate *)time
                    timeZone:(NSTimeZone *)timeZone
{
  self = [super initWithFrame:CGRectMake(0, 0, 320, 216)];
  if (self) {
    self.includeTimeType = NO;
    
    [self commonInit];
    
    if (! time) {
      time = [NSDate date];
    }
    self.timePicker.datePickerMode = UIDatePickerModeDate;
    self.timePicker.timeZone = timeZone;
    self.timePicker.date     = time;
    self.timePicker.minimumDate = nil;
    self.timePicker.maximumDate = nil;
  }
  return self;
}

- (void)setTimePicker:(UIDatePicker *)timePicker {
  if (timePicker != nil) {
    _timePicker = timePicker;
  }
}

- (NSDate *)selectedDate
{
  return self.timePicker.date;
}

- (TKTimeType)selectedTimeType
{
  if (! self.includeTimeType)
    return TKTimeTypeNone;
  
  if (self.timeTypeSelector.selectedSegmentIndex == 0)
    return TKTimeTypeLeaveASAP;
  
  if (self.timeTypeSelector.selectedSegmentIndex == 1)
    return TKTimeTypeLeaveAfter;
  
  if (self.timeTypeSelector.selectedSegmentIndex == 2)
    return TKTimeTypeArriveBefore;
  
  return TKTimeTypeNone;
}

#pragma mark - User Interaction

- (IBAction)doneButtonPressed:(id)sender
{
  self.doneSelector.selectedSegmentIndex = -1;
  
	if ([self isBeingOverlaid]) {
		[self tappedOverlay:sender];
	} else if ([self.delegate respondsToSelector:@selector(timePickerRequestsResign:)]) {
		[self.delegate timePickerRequestsResign:self];
  }
}


- (IBAction)nowButtonPressed:(id)sender
{
  if (self.selectAction) {
    self.selectAction(TKTimeTypeLeaveASAP, [NSDate date]);
    self.selectAction = nil;
  }

	if ([self isBeingOverlaid]) {
		[self tappedOverlay:sender];
	} else if ([self.delegate respondsToSelector:@selector(timePickerRequestsResign:)]) {
		[self.delegate timePickerRequestsResign:self];
	} else {
    [self.timePicker setDate:[NSDate date] animated:YES];
  }
}

- (IBAction)timePickerChanged:(id)sender
{
#pragma unused(sender)
  self.didSetTime = YES;
  if (TKTimeTypeLeaveASAP == [self selectedTimeType]) {
    [self setSelectedTimeType:TKTimeTypeLeaveAfter];
  } else {
    [self setSelectedTimeType:[self selectedTimeType]];
  }
}

- (IBAction)timeSelectorChanged:(id)sender
{
  if (TKTimeTypeLeaveASAP == [self selectedTimeType]) {
    [self nowButtonPressed:sender];
  } else {
    [self setSelectedTimeType:[self selectedTimeType]];
  }
}

- (void)tappedOverlay:(id)sender
{
  if (self.selectAction && self.didSetTime) {
    self.selectAction([self selectedTimeType], self.timePicker.date);
    self.selectAction = nil;
  }
  
  [super tappedOverlay:sender];
}


#pragma mark - Private methods

- (void)commonInit
{
  static const CGFloat kButtonHeight = 44;
  
  self.backgroundColor = UIColor.tkBackground;
  
  UIDatePicker *timePicker = [[UIDatePicker alloc] init];
  timePicker.datePickerMode = UIDatePickerModeDateAndTime;
  if (@available(iOS 14, *)) {
    timePicker.preferredDatePickerStyle = UIDatePickerStyleWheels;
    [timePicker sizeToFit];
  }
  timePicker.minimumDate = [NSDate dateWithTimeIntervalSinceNow:60 * 60 * 24 * -31]; // 1 month ago
  timePicker.maximumDate = [NSDate dateWithTimeIntervalSinceNow:60 * 60 * 24 * 31]; // 1 month
  timePicker.locale = [TKStyleManager applicationLocale]; // Set the 24h setting
  timePicker.autoresizingMask = UIViewAutoresizingFlexibleWidth;
  [timePicker addTarget:self
                 action:@selector(timePickerChanged:)
       forControlEvents:UIControlEventValueChanged];
  [self addSubview:timePicker];
  self.timePicker = timePicker;
  
  CGRect timePickerFrame = timePicker.frame;
  timePickerFrame.size.width = CGRectGetWidth(self.frame);
  timePickerFrame.origin.y = kButtonHeight;
  self.timePicker.frame = timePickerFrame;

  UISegmentedControl *timeTypeSelector;
  if (self.includeTimeType) {
     timeTypeSelector = [[UISegmentedControl alloc] initWithItems:@[
                                                                    Loc.Now,
                                                                    Loc.LeaveAt,
                                                                    Loc.ArriveBy]];
    [timeTypeSelector addTarget:self
                         action:@selector(timeSelectorChanged:)
               forControlEvents:UIControlEventValueChanged];
    self.timeTypeSelector = timeTypeSelector;
  }
  
  UISegmentedControl *doneSelector = [[UISegmentedControl alloc] initWithItems:@[Loc.Done]];
  [doneSelector addTarget:self
                  action:@selector(doneButtonPressed:)
        forControlEvents:UIControlEventValueChanged];
  self.doneSelector = doneSelector;

  UIToolbar *toolbar = [[UIToolbar alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(timePickerFrame), kButtonHeight)];
  toolbar.autoresizingMask = UIViewAutoresizingFlexibleWidth;
  
  NSMutableArray *items = [NSMutableArray array];
  if (timeTypeSelector) {
    [items addObject:[[UIBarButtonItem alloc] initWithCustomView:timeTypeSelector]];
  }
  [items addObject:[[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil]];
  [items addObject:[[UIBarButtonItem alloc] initWithCustomView:doneSelector]];
  toolbar.items = items;
  
  [toolbar setBackgroundImage:[UIImage new]
           forToolbarPosition:UIToolbarPositionAny
                   barMetrics:UIBarMetricsDefault];
  toolbar.backgroundColor = self.backgroundColor;
  [self addSubview:toolbar];

  self.frame = CGRectMake(0, 0, CGRectGetWidth(timePickerFrame), CGRectGetHeight(timePickerFrame) + kButtonHeight);
}

- (void)setSelectedTimeType:(TKTimeType)timeType
{
  switch (timeType) {
    case TKTimeTypeLeaveASAP:
      self.timeTypeSelector.selectedSegmentIndex = 0;
      [self.timePicker setDate:[NSDate date] animated:YES];
      break;
      
    case TKTimeTypeLeaveAfter: {
      self.timeTypeSelector.selectedSegmentIndex = 1;
      break;
    }
      
    case TKTimeTypeArriveBefore: {
      self.timeTypeSelector.selectedSegmentIndex = 2;
      break;
    }
      
    default:
      break;
  }
  
  self.didSetTime = YES;
  if ([self.delegate respondsToSelector:@selector(timePicker:pickedDate:forType:)]) {
    [self.delegate timePicker:self
                   pickedDate:self.timePicker.date
                      forType:timeType];
  }
}

@end

File: TripKitUI/views/results/TKUITimePickerSheet.swift
//
//  TKUITimePickerSheet.swift
//  TripKitUI-iOS
//
//  Created by Adrian Schönig on 31/1/21.
//  Copyright © 2021 SkedGo Pty Ltd. All rights reserved.
//

import UIKit

import TripKit

@objc
public protocol TKUITimePickerSheetDelegate: class {

  @objc(timePicker:pickedDate:forType:)
  func timePicker(_ picker: TKUITimePickerSheet, pickedDate: Date, for type: TKTimeType)

  @objc
  func timePickerRequestsResign(_ picker: TKUITimePickerSheet)

}

public class TKUITimePickerSheet: TKUISheet {
  
  public var selectAction: (TKTimeType, Date) -> Void = { _, _ in }
  
  public weak var delegate: TKUITimePickerSheetDelegate?
  
  public var selectedDate: Date {
    get {
      timePicker.date
    }
    set {
      timePicker.date = newValue
    }
  }
  
  public var selectedTimeType: TKTimeType {
    get {
      switch timeTypeSelector?.selectedSegmentIndex {
      case 0: return .leaveASAP
      case 1: return .leaveAfter
      case 2: return .arriveBefore
      default: return .none
      }
    }
    set {
      switch newValue {
      case .leaveASAP:
        timeTypeSelector?.selectedSegmentIndex = 0
        timePicker.setDate(.init(), animated: true)
      case .leaveAfter:
        timeTypeSelector?.selectedSegmentIndex = 1
      case .arriveBefore:
        timeTypeSelector?.selectedSegmentIndex = 2
      case .none:
        break
      }
      
      didSetTime = true
      delegate?.timePicker(self, pickedDate: selectedDate, for: newValue)
    }
  }
  
  private var didSetTime: Bool
  private weak var timePicker: UIDatePicker!
  private weak var timeTypeSelector: UISegmentedControl!
  private weak var doneSelector: UISegmentedControl!

  public convenience init(date: Date, timeZone: TimeZone) {
    self.init(date: date, showTime: false, timeType: .none, timeZone: timeZone)
  }
  
  public convenience init(time: Date, timeType: TKTimeType = .none, timeZone: TimeZone) {
    self.init(date: time, showTime: true, timeType: timeType, timeZone: timeZone)
  }
  
  private init(date: Date, showTime: Bool, timeType: TKTimeType, timeZone: TimeZone) {
    didSetTime = false

    super.init(frame: .init(origin: .zero, size: .init(width: 320, height: 116)))
    
    overlayColor = .tkSheetOverlay
    backgroundColor = .tkBackground
    
    let timePicker = UIDatePicker()
    timePicker.datePickerMode = showTime ? .dateAndTime : .date
    timePicker.date = date
    timePicker.timeZone = timeZone
    
    if #available(iOS 13.4, *) {
      timePicker.preferredDatePickerStyle = .wheels
      timePicker.sizeToFit()
    }

    if showTime {
      // Limit to a month ago until one month from now
      timePicker.minimumDate = .init(timeIntervalSinceNow: 60 * 60 * 24 * -31)
      timePicker.maximumDate = .init(timeIntervalSinceNow: 60 * 60 * 24 * 31)
    }
    
    timePicker.locale = TKStyleManager.applicationLocale() // set 24h setting
    timePicker.autoresizingMask = .flexibleWidth
    timePicker.addTarget(self, action: #selector(timePickerChanged(sender:)), for: .valueChanged)
    self.addSubview(timePicker)
    self.timePicker = timePicker
    
    let selector: UISegmentedControl! // as `timeTypeSelector` is weak
    if timeType != .none {
      selector = UISegmentedControl(items: [Loc.Now, Loc.LeaveAt, Loc.ArriveBy])
      selector.addTarget(self, action: #selector(timeSelectorChanged(sender:)), for: .valueChanged)
      self.timeTypeSelector = selector
      
      // this sets the selected section index
      self.selectedTimeType = timeType
    }
    
    // Yes, a segmented control with one element. This is for consistent styling.
    let doneSelector = UISegmentedControl(items: [Loc.Done])
    doneSelector.addTarget(self, action: #selector(doneButtonPressed(sender:)), for: .valueChanged)
    self.doneSelector = doneSelector
    
    let toolbar = UIToolbar(frame: .init(x: 0, y: 0, width: timePicker.frame.width, height: 44))
    toolbar.autoresizingMask = .flexibleWidth
    toolbar.items = [
      timeTypeSelector.map(UIBarButtonItem.init(customView:)),
      UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil),
      UIBarButtonItem(customView: doneSelector)
    ].compactMap { $0 }
    toolbar.setBackgroundImage(.init(), forToolbarPosition: .any, barMetrics: .default)
    toolbar.backgroundColor = self.backgroundColor
    addSubview(toolbar)
    
    self.frame = .init(x: 0, y: 0, width: timePicker.frame.width, height: timePicker.frame.height + toolbar.frame.height)
  }
  
  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
  
  public override func tappedOverlay(_ sender: Any!) {
    if didSetTime {
      selectAction(selectedTimeType, selectedDate)
      selectAction = { _, _ in }
    }
    
    super.tappedOverlay(sender)
  }
  
  // MARK: Actions
  
  @objc
  func timePickerChanged(sender: Any) {
    didSetTime = true
    
    if selectedTimeType == .leaveASAP {
      selectedTimeType = .leaveAfter
    }
  }
  
  @objc
  func timeSelectorChanged(sender: Any) {
    if selectedTimeType == .leaveASAP {
      nowButtonPressed(sender: sender)
    }
  }

  @objc
  func nowButtonPressed(sender: Any) {
    selectAction(.leaveASAP, .init())
    selectAction = { _, _ in }
    
    if isBeingOverlaid() {
      tappedOverlay(sender)
    } else if let delegate = delegate {
      delegate.timePickerRequestsResign(self)
    } else {
      timePicker.setDate(.init(), animated: true)
    }
  }

  @objc
  func doneButtonPressed(sender: Any) {
    doneSelector.selectedSegmentIndex = -1
    
    if isBeingOverlaid() {
      tappedOverlay(sender)
    } else if let delegate = delegate {
      delegate.timePickerRequestsResign(self)
    } else {
      assertionFailure("Done pressed, but don't know what to do. Set a delegate?")
    }
  }
  
}
